### DISCUSSION TITLE
Splitting configuration files

### DISCUSSION THREAD

**Discussion by God-damnit-all** [wezterm/wezterm#914] [Q&A]
I was thinking it would be nice to split off some of the settings in my `wezterm.lua` into separate files. I know you can create lua modules with their own functions, but it seems everything related to the Config struct type is limited to `wezterm.lua`... as far as I can tell, anyway.

This would be especially useful for creating alternate configurations that would reuse some of the same settings. So fontconfig.lua could be loaded by `wezterm.lua` and by `altconfig.lua` (when using `wezterm --config-file "%WEZTERM_CONFIG_DIR%\altconfig.lua"`). This makes it so you'd only have to edit fontconfig.lua instead of making the same changes to two separate files.

But, I know there's a lot of undocumented features, so it's probably possible to do this already. Question is, how? And, if it is, would it be possible to have these 'subconfig' files monitored for changes for a hot reload just like the main config file?

**Comment by bew** [comment]
For configs in multi files, take a look at my configs if you want:
https://github.com/bew/dotfiles/tree/main/gui/wezterm
I've split the configs by topic, each module returns a subset of the config, and all the sub-configs are merged and returned for wezterm.
I have a few functions to help, in [`lib/mystdlib.lua`](https://github.com/bew/dotfiles/blob/main/gui/wezterm/lib/mystdlib.lua)

The other files are not tracked for auto-reload, but I have disabled this feature, I don't like it `¯\_(ツ)_/¯`

**Reply by God-damnit-all** [reply]
This looks pretty good! Shame about no auto-reload, but everything is really organized.

I'm surprised it's this complicated to do multi files for the config, this must've been a real hassle to set up.

**Reply by bew** [reply]
Thanks! It's not really a hassle for me, I know (and like!) Lua, and I like to have things organized in my own way, like this (: 

About the complexity, you know, at the end of the day some code will need to do this logic of merging various parts of the config. Wezterm is kept simple as it only require you to return a Lua table with specific fields for the config, how to create that config table is up to the config file, so if you want to split, you'll have to make the merges yourself.
I don't think there really are other ways to do this.. (unless you have a suggestion?)

---
For 'free' auto-reload, maybe there could be a custom function like `wezterm.require_and_track("some_file")`, that would require a file and mark it for tracking for the auto-reload system, and then you'd use that instead of `require`.
(or maybe not replace require, but have a way to track a file / the current file?)
@wez what do you think of this kind of thing?

**Reply by wez** [reply]
I think adding a custom searcher (https://www.lua.org/manual/5.3/manual.html#pdf-package.searchers) ahead of the default second searcher, that uses [package.searchpath](https://www.lua.org/manual/5.3/manual.html#pdf-package.searchpath) to determine the path, save it in a global, and then calls that default second searcher to actually load the module would give wezterm a list of paths to watch.

**Reply by wez** [reply]
ad42f1d7c4707715539cd9ed10248b66b85ab82f implements auto reload for required files as well
